File: HW 2 (Java)/src/Quadrilateral.java
import java.util.Arrays;
import java.util.List;

public class Quadrilateral implements Positionable, TwoDShape {


    private final TwoDPoint[] vertices = new TwoDPoint[4];


    public Quadrilateral(double... vertices) {
        this(TwoDPoint.ofDoubles(vertices));
    }

    public Quadrilateral(List<TwoDPoint> vertices) {
        int n = 0;
        for (TwoDPoint p : vertices) this.vertices[n++] = p;
        if (!isMember(vertices))
            throw new IllegalArgumentException(String.format("Invalid set of vertices specified for %s",
                    this.getClass().getCanonicalName()));
    }


    /**
     * Given a list of four points, adds them as the four vertices of this quadrilateral in the order provided in the
     * list. This is expected to be a counterclockwise order of the four corners.
     *
     * @param points the specified list of points.
     * @throws IllegalStateException if the number of vertices provided as input is not equal to four.
     */
    @Override
    // should we throw an error when given arrayList is not consisted of 4 twoDPoint,

    public void setPosition(List<? extends Point> points) {
        if (points.size() != 4)
            throw new IllegalStateException();

        for (int i = 0; i < 4; i++) {
            if (!(points.get(i) instanceof TwoDPoint))
                throw new IllegalArgumentException();
        }

        vertices[0] = (TwoDPoint) points.get(0);
        vertices[1] = (TwoDPoint) points.get(1);
        vertices[2] = (TwoDPoint) points.get(2);
        vertices[3] = (TwoDPoint) points.get(3);
    }

    @Override
    public List<TwoDPoint> getPosition() {
        return Arrays.asList(vertices);
    }

    /**
     * @return the lengths of the four sides of the quadrilateral. Since the setter {@link Quadrilateral#setPosition(List)}
     *         expected the corners to be provided in a counterclockwise order, the side lengths are expected to be in
     *         that same order.
     */
    protected double[] getSideLengths() {
        double [] sideArr = new double [4];
        double side1 = Math.pow(Math.pow(vertices[0].getX() - vertices[1].getX(), 2) + Math.pow(vertices[0].getY() - vertices[1].getY(), 2), 0.5);
        double side1Round = (Math.round(side1 * 1000.0) / 1000.0);
        double side2 = Math.pow(Math.pow(vertices[1].getX() - vertices[2].getX(), 2) + Math.pow(vertices[1].getY() - vertices[2].getY(), 2), 0.5);
        double side2Round = (Math.round(side2 * 1000.0) / 1000.0);
        double side3 = Math.pow(Math.pow(vertices[2].getX() - vertices[3].getX(), 2) + Math.pow(vertices[2].getY() - vertices[3].getY(), 2), 0.5);
        double side3Round = (Math.round(side3 * 1000.0) / 1000.0);
        double side4 = Math.pow(Math.pow(vertices[3].getX() - vertices[0].getX(), 2) + Math.pow(vertices[3].getY() - vertices[0].getY(), 2), 0.5);
        double side4Round = (Math.round(side4 * 1000.0) / 1000.0);
        sideArr[0] = side1Round;
        sideArr[1] = side2Round;
        sideArr[2] = side3Round;
        sideArr[3] = side4Round;
        return sideArr; // TODO
    }

    @Override
    public int numSides() { return 4; }

    @Override
    public boolean isMember(List<? extends Point> vertices) { return vertices.size() == 4; }

    public double area() {
        double area = this.getSideLengths()[0] * this.getSideLengths()[1];
        return area; // TODO
    }


}

File: HW 3 (Python)/hw3-codebase/quadrilateral.py


from two_d_point import TwoDPoint



class Quadrilateral:

    def __init__(self, *floats):
        if not (len(list(floats)) == 8):
            raise TypeError("A {} cannot be formed by the given coordinates.".format(self.__class__.__name__))
        points = TwoDPoint.from_coordinates(list(floats))
        self.__vertices = tuple(points[0:4])
        if ((self.vertices[0] == self.vertices[1]) or (self.vertices[0] == self.vertices[2]) or (self.vertices[0] == self.vertices[3]) or (self.vertices[1] == self.vertices[2]) or (self.vertices[1] == self.vertices[3]) or (self.vertices[2] == self.vertices[3])):
            raise TypeError("A {} cannot be formed by the given coordinates.".format(self.__class__.__name__))


    @property
    def vertices(self):
        return self.__vertices

    def side_lengths(self):
        side1 = pow(pow(self.__vertices[1].x - self.__vertices[0].x , 2) + pow(self.__vertices[1].y - self.__vertices[0].y , 2),0.5)
        side1 = round(side1,5)
        side2 = pow(pow(self.__vertices[0].x - self.__vertices[3].x , 2) + pow(self.__vertices[0].y - self.__vertices[3].y , 2),0.5)
        side2 = round(side2,5)
        side3 = pow(pow(self.__vertices[3].x - self.__vertices[2].x , 2) + pow(self.__vertices[3].y - self.__vertices[2].y , 2),0.5)
        side3 = round(side3,5)
        side4 = pow(pow(self.__vertices[2].x - self.__vertices[1].x , 2) + pow(self.__vertices[2].y - self.__vertices[1].y , 2),0.5)
        side4 = round(side4,5)
        return side1, side2, side3, side4  # TODO

    def smallest_x(self):
        lst = [self.vertices[0].x, self.vertices[1].x, self.vertices[2].x, self.vertices[3].x]
        min_num = min(lst)
        return min_num
            # TODO

    def __str__(self) -> str:
        return 'I am a Quadrilateral with ' + str(self.vertices[0]) + " , "+ str(self.vertices[1]) + " , " + str(self.vertices[2]) + " , " + str(self.vertices[3])

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Quadrilateral):
            if ( (self.vertices[0] == other.vertices[0]) and (self.vertices[1] == other.vertices[1]) and (self.vertices[2] == other.vertices[2]) and (self.vertices[3] == other.vertices[3])):
                return True
        return False  # TODO

    def __ne__(self, other: object) -> bool:
        return not self.__eq__(other)











